### ISSUE TITLE
Learning rate-free coin betting optimizer

### ISSUE THREAD

**Issue by albcab** [google-deepmind/optax#530]
I have a working code for the backpropagating version of the COntinuous COin Betting (COCOB) optimizer, see [Algorithm 2](https://arxiv.org/pdf/1705.07795.pdf). Is there any interest in having it in optax? I can do the contribution.

I am using it to do learning rate-free SVGD, see [Algorithm 6](https://arxiv.org/pdf/2301.11294.pdf), in case you are interested in useful applications. 

The only caveat is that new parameters are updated from the initial parameters and not from the current parameters. So the returned `update` need to be `params - (init_params + gradient_ascent_update)` so that `optax.apply_updates` with these updates returns `init_params + gradient_ascent_update` (line 10 of [Algorithm 2](https://arxiv.org/pdf/1705.07795.pdf)). I've found this to create a smaller floating point precision error than making `update` equal to `previous_step_gradient_ascent_update - gradient_ascent_update` (since `params - previous_step_gradient_ascent_update` equals `init_params`).

**Comment by mtthss** [comment]
That sound very interesting, would be interested in implementing it and upstreaming to optax/contrib?

**Comment by fabianp** [comment]
Closing as this is now part of contrib
